pkg/webhooks/mutator/virt-launcher-mutator_test.go

```diff
@@ -0,0 +1,159 @@
+package mutator
+
+import (
+	"fmt"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/rand"
+
+	"github.com/kubevirt/hyperconverged-cluster-operator/api/v1beta1"
+	"github.com/kubevirt/hyperconverged-cluster-operator/controllers/commonTestUtils"
+	"github.com/kubevirt/hyperconverged-cluster-operator/pkg/util"
+)
+
+const (
+	hcoNamespace = HcoValidNamespace
+	podNamespace = "fake-namespace"
+)
+
+var _ = Describe("virt-launcher webhook mutator", func() {
+	mp := resource.MustParse
+
+	Describe("resource multiplier", func() {
+		DescribeTable("produces correct results", func(input, expectedOutput resource.Quantity, ratio float64) {
+			mutator := getVirtLauncherMutator()
+			actualOutput := mutator.multiplyResource(input, ratio)
+			Expect(actualOutput.Equal(expectedOutput)).To(BeTrue(), fmt.Sprintf("expected %s to equal %s", actualOutput.String(), expectedOutput.String()))
+		},
+			Entry("CPU: 100m with ratio 2", mp("100m"), mp("200m"), 2.0),
+			Entry("CPU: 700m with ratio 2", mp("700m"), mp("1400m"), 2.0),
+			Entry("CPU: 2.4 with ratio 2", mp("2.4"), mp("4800m"), 2.0),
+			Entry("CPU: 0.4 with ratio 2", mp("0.4"), mp("800m"), 2.0),
+			Entry("CPU: 200m with ratio 0.5", mp("200m"), mp("100m"), 0.5),
+			Entry("CPU: 1 with ratio 0.5", mp("1"), mp("500m"), 0.5),
+
+			Entry("Memory: 256 with ratio 3.0", mp("256"), mp("768"), 3.0),
+			Entry("Memory: 256M with ratio 3.0", mp("256M"), mp("768M"), 3.0),
+			Entry("Memory: 256Mi with ratio 3.0", mp("256Mi"), mp("768Mi"), 3.0),
+			Entry("Memory: 256Gi with ratio 3.0", mp("256Gi"), mp("768Gi"), 3.0),
+			Entry("Memory: 700M with ratio 3.0", mp("700M"), mp("2100M"), 3.0),
+			Entry("Memory: 256M with ratio 3.0", mp("260M"), mp("52M"), 0.2),
+		)
+
+	})
+
+	DescribeTable("set resource ratio", func(memRatio, cpuRatio string, podResources, expectedResources v1.ResourceRequirements) {
+		mutator := getVirtLauncherMutator()
+		launcherPod := getFakeLauncherPod()
+		hco := getHco()
+		hco.Annotations = map[string]string{
+			cpuLimitToRequestRatioAnnotation:    cpuRatio,
+			memoryLimitToRequestRatioAnnotation: memRatio,
+		}
+
+		launcherPod.Spec.Containers[0].Resources = podResources
+		err := mutator.handleVirtLauncherCreation(launcherPod, hco)
+		Expect(err).ToNot(HaveOccurred())
+
+		resources := launcherPod.Spec.Containers[0].Resources
+		Expect(resources.Limits[v1.ResourceCPU].Equal(expectedResources.Limits[v1.ResourceCPU])).To(BeTrue())
+		Expect(resources.Requests[v1.ResourceCPU].Equal(expectedResources.Requests[v1.ResourceCPU])).To(BeTrue())
+		Expect(resources.Limits[v1.ResourceMemory].Equal(expectedResources.Limits[v1.ResourceMemory])).To(BeTrue())
+		Expect(resources.Requests[v1.ResourceMemory].Equal(expectedResources.Requests[v1.ResourceMemory])).To(BeTrue())
```

Comment:
Isn't the gomega's `Equals` matcher work here?

Reply:
Unfortunately not :(
With Gomega's Equals I get errors that are similar to:
```
  [FAILED] Expected
      <resource.Quantity>: {
          i: {value: 150000000000, scale: -3},
          d: {Dec: nil},
          s: "150M",
          Format: "DecimalSI",
      }
  to equal
      <resource.Quantity>: {
          i: {value: 150, scale: 6},
          d: {Dec: nil},
          s: "150M",
          Format: "DecimalSI",
      }
```

Although the two quantities are effectively equal (to `150M`).